### PR TITLE
Update uDiscovery specs 

### PR DESCRIPTION
**Pull request by mayureshkar**
This PR is to make updates to uDiscovery specifications to make APIs more intent revealing and to removed unwanted complexities.

#137

### CODE REVIEW

File: up-l3/udiscovery/v3/FindUE.puml
Comment (stevenhartley) [inline up-l3/udiscovery/v3/FindUE.puml:1]: Personally, I think it would be simpler to just state that lookups traverse from local to domain to centeral in lieu of the complicated diagrams to simplify the specs.

File: up-l3/udiscovery/v3/README.adoc [hunk 1]
```diff
@@ -31,9 +30,9 @@ SPDX-License-Identifier: Apache-2.0
...
+uE's (Software Entities) will be deployed dynamically on devices that are scattered throughout the uProtocol network. In order for these  uE's to know about what is available on their own device or on other devices, we need the ability to discover uEs, their locations and what they serve (resources, methods, messages, etc...). There are number of Internet protocols such as DNS, Zeroconf, and others, to discover devices and services running on said devices. Intent of uDiscovery is to reuse as much as possible, existing Internet standards for discovering uThings as well.
```
Comment (stevenhartley) [inline up-l3/udiscovery/v3/README.adoc:33]:
```suggestion
uE's (Software Entities) will be deployed dynamically on devices that are scattered throughout the uProtocol network. In order for these uE's to know about what is available on their own device or on other devices, we need the ability to discover uEs, their locations and what they serve (resources, methods, messages, etc...). There are number of Internet protocols such as DNS, Zeroconf, and others, to discover devices and services running on said devices. Intent of uDiscovery is to reuse as much as possible, existing Internet standards for discovering uThings as well.
```

File: up-l3/udiscovery/v3/README.adoc [hunk 2]
```diff
@@ -45,15 +44,16 @@ The purpose of this proposal is to define the discovery architecture so that uEs
 |Term |Definition
 |DNS |Domain Name System covered in various IETF RFCs
 |CDS |Central Discovery Service implementation of uDiscovery
+|DDS | Domain Discovery Service(local to a domain) implemenation of uDiscovery
```
Comment (stevenhartley) [inline up-l3/udiscovery/v3/README.adoc:47]:
```suggestion
|uDDS | Domain Discovery Service(local to a domain) implementation of uDiscovery
```

File: up-l3/udiscovery/v3/README.adoc [hunk 3]
```diff
@@ -45,15 +44,16 @@ The purpose of this proposal is to define the discovery architecture so that uEs
 |Term |Definition
 |DNS |Domain Name System covered in various IETF RFCs
 |CDS |Central Discovery Service implementation of uDiscovery
```
Comment (stevenhartley) [inline up-l3/udiscovery/v3/README.adoc:46]:
```suggestion
|uCDS |Central Discovery Service implementation of uDiscovery
```

File: up-l3/udiscovery/v3/README.adoc [hunk 4]
```diff
@@ -45,15 +44,16 @@ The purpose of this proposal is to define the discovery architecture so that uEs
...
+|DDS | Domain Discovery Service(local to a domain) implemenation of uDiscovery
 |LDS |Local (to a device) implementation of uDiscovery
```
Comment (stevenhartley) [inline up-l3/udiscovery/v3/README.adoc:48]:
```suggestion
|uLDS |Local (to a device) implementation of uDiscovery
```

File: up-l3/udiscovery/v3/README.adoc [hunk 5]
```diff
@@ -45,15 +44,16 @@ The purpose of this proposal is to define the discovery architecture so that uEs
...
-|Local Node |Node who is local to the device (local information)
-|Remote Node |Node that is fetched from the CDS and cached in the LDS
+|Local Entity |Entity  who is local to the device (local information)
```
Comment (stevenhartley) [inline up-l3/udiscovery/v3/README.adoc:49]: Remove the remote and local entity rows in the table

File: up-l3/udiscovery/v3/README.adoc [hunk 6]
```diff
@@ -70,30 +70,28 @@ NOTE: Components described in <<udiscovery-ues>> below all implement link:../../
 |Component |Description |Requirements
 
 |*Local uDiscovery Service (LDS)*
```
Comment (stevenhartley) [inline up-l3/udiscovery/v3/README.adoc:72]:
```suggestion
|*Local uDiscovery Service (uLDS)*
```

File: up-l3/udiscovery/v3/README.adoc [hunk 7]
```diff
@@ -70,30 +70,28 @@ NOTE: Components described in <<udiscovery-ues>> below all implement link:../../
...
-|*Central uDiscovery Service (CDS)*
+|*Central uDiscovery Service (uCDS)*
 |Central repository storing all deployed uDevice information in the network of connected devices residing in the Cloud. It is a superset of the local uDiscovery database content
 a|
 * *MUST* have one CDS that is globally addressable by the various domain/device LDS
 |===
 
-Function specific requirements shall be covered later in this document.
-
 === Hierarchical Specifications
```
Comment (stevenhartley) [inline up-l3/udiscovery/v3/README.adoc:92]: This section is no longer relavent and should be removed IMHO

File: up-l3/udiscovery/v3/README.adoc [hunk 8]
```diff
@@ -206,7 +204,7 @@ NOTE: It is *STRONGLY RECOMMENDED* to scope your property names to avoid namespa
 
 === Database Access
```
Comment (stevenhartley) [inline up-l3/udiscovery/v3/README.adoc:205]: This should not be in the specification as it is an implementation detail that is not part of the API specifications (more on implementation detail for a given DS. Please remove, same with policy. I don't think we should put database specific details here. It could be a best practice section somewhere else.

File: up-l3/udiscovery/v3/README.adoc [hunk 9]
```diff
@@ -254,107 +251,29 @@ API requirements related to Node metadata shall be covered in the subsequent sec
 
 In the following section we will explain the various APIs and interfaces that are defined in uDiscovery and their requirements. Interface definitions (input and output parameters, etc...) are covered in the link:../../../up-core-api/uprotocol/core/udiscovery/v3/udiscovery.proto[udiscovery.proto].
 
-=== Notifications
```
Comment (stevenhartley) [inline up-l3/udiscovery/v3/README.adoc:257]: No more notifications? If uEs are added or removed, how are we notified?

File: up-l3/udiscovery/v3/README.adoc [hunk 10]
```diff
@@ -254,107 +251,29 @@ API requirements related to Node metadata shall be covered in the subsequent sec
...
-image:notifications.svg[Notifications]
-
 ===  Read/Query APIs
 
 Query APIs are used to lookup content in the database, either to resolve URIs (to be used by applications) or to fetch content of a database.
 
 * Any uE *MAY* call the query APIs defined in the sections below
-* *MUST NOT* return Nodes that are flagged as `inactive`
-* Remote Nodes that have `expired` *MUST* be refreshed to the CDS
+* Remote Nodes that are `expired` *MUST* be refreshed to the CDS
```
Comment (stevenhartley) [inline up-l3/udiscovery/v3/README.adoc:259]: All node references need to be removed

File: up-l3/udiscovery/v3/README.adoc [hunk 11]
```diff
@@ -254,107 +251,29 @@ API requirements related to Node metadata shall be covered in the subsequent sec
...
-.Delete Node
-image:delete_node.svg[Delete Node]
-
-
-==== ResolveUri
-
-`ResolveUri()` is used to lookup names from ids or vice versa meaning to go to/from Long Uri from/to micro Uri. 
-
-For portability between SDKs, Long Form URIs shall be serialized to String and Short form Uris shall be serialized to Bytes per link:../../../basics/uri.adoc[URI Specifications]. 
+* *MUST* update `last_accessed` Node attribute when API is called
```
Comment (stevenhartley) [inline up-l3/udiscovery/v3/README.adoc:273]: No longer relevant for the API.

File: up-l3/udiscovery/v3/README.adoc [hunk 12]
```diff
@@ -254,107 +251,29 @@ API requirements related to Node metadata shall be covered in the subsequent sec
...
-
-==== Updating a Node
-
-Below is the sequence when a change happens in the database
-.Updating Nodes
-image:update_nodes.svg["Update Nodes"]
+image:FindUE.svg[FinduE]
 
-==== Adding Node(s)
+==== Get UE Topics
```
Comment (stevenhartley) [inline up-l3/udiscovery/v3/README.adoc:269]: Please add more details as to why this APi exists, who would use it when

File: up-l3/udiscovery/v3/README.adoc [hunk 13]
```diff
@@ -254,107 +251,29 @@ API requirements related to Node metadata shall be covered in the subsequent sec
...
-For portability between SDKs, Long Form URIs shall be serialized to String and Short form Uris shall be serialized to Bytes per link:../../../basics/uri.adoc[URI Specifications]. 
+* *MUST* update `last_accessed` Node attribute when API is called
 
- * *MUST* be passed either a Long or Short form URI and return both the Micro and Long form URI if successful
 
 
 == Failure Recovery
 
-In the event that the databases between the CDS and LDS becomes out of sync, the discovery service components (uLDS, uDDS, uCDS) *MAY* fetch the contents using `FindNodes()` API.
+In the event that the databases between the CDS and LDS becomes out of sync, the discovery service components (uLDS, uDDS, uCDS) *MAY* fetch the contents using read APIs.
```
Comment (stevenhartley) [inline up-l3/udiscovery/v3/README.adoc:279]: Database specific detail that is out of scope for this API documentation.

File: up-l3/udiscovery/v3/README.adoc
Comment (stevenhartley) [inline up-l3/udiscovery/v3/README.adoc:1]: the other new APIs in the proto are not covered here and there is no explanation about their purpose.

**Review by stevenhartley — CHANGES_REQUESTED** [review]
Please see feedback

File: up-core-api/uprotocol/core/udiscovery/v3/udiscovery.proto [hunk 14]
```diff
@@ -58,74 +49,27 @@ service uDiscovery {
...
-  // 3. uService: `//device.domain/body.access/`
-  // 4. uResource: `//device.domain/body.access//door.door_front_left`
-  // **NOTE:** You MUST have read permission to the node in the database
-  rpc FindNodes(FindNodesRequest) returns (FindNodesResponse) {
-    option (method_id) = 3;
-  }
-
-  // Query the database to fetch a list of 1 or more properties for a given node.
-  rpc FindNodeProperties(FindNodePropertiesRequest) returns (FindNodePropertiesResponse) {
+  // Retrieve the list of uEs under a device.
```
[the remark this answers is not in the document]
Comment (mayureshkar) [reply up-core-api/uprotocol/core/udiscovery/v3/udiscovery.proto:61]: GetuETopics could be used to fetch the list of topics serviced by a given uE. It is a focused API to give that information as opposed to FindUE which gives everything about uE.

File: up-core-api/uprotocol/core/udiscovery/v3/udiscovery.proto [hunk 15]
```diff
@@ -139,233 +83,80 @@ service uDiscovery {
...
-message DeleteNodesRequest {
-  repeated string uris = 1;   // uProtocol formatted URI
-}
-
-
-// FindNodesRequest passed to FindNodes()
-message FindNodesRequest {
-  // uProtocol formatted URI for the node to search, ex. '//vcu.VIN/core.app.hartley'
-  // shall return the 'core.app.hartley' node
+message FinduERequest {
```
[the remark this answers is not in the document]
Comment (mayureshkar) [reply up-core-api/uprotocol/core/udiscovery/v3/udiscovery.proto:86]: One would use this to get a list of qulified URIs. How do we mandate input to be a URI. example: I want to find instances for service1

File: up-core-api/uprotocol/core/udiscovery/v3/udiscovery.proto [hunk 16]
```diff
@@ -41,15 +41,6 @@ service uDiscovery {
   option (version_minor) = 0;
   option (id) = 1;
 
-  // uDiscovery Node Change Notification that sends the Update message
```
[the remark this answers is not in the document]
Comment (mayureshkar) [reply up-core-api/uprotocol/core/udiscovery/v3/udiscovery.proto:44]: Added it back in. Had originally planned to add it back in once we figure out notifications.

File: up-core-api/uprotocol/core/udiscovery/v3/udiscovery.proto [hunk 17]
```diff
@@ -58,74 +49,27 @@ service uDiscovery {
...
-  rpc LookupUri(v1.UUri) returns (LookupUriResponse) {
+  rpc FinduE(FinduERequest) returns (FinduEResponse) {
     option (method_id) = 1;
   }
 
-  // Update a node in the database.
-  // What is returned is the status of the API request as uprotocol.v1.UStatus.
-  // **NOTE:** You MUST have write permission to the node in the database
-  rpc UpdateNode(UpdateNodeRequest) returns (uprotocol.v1.UStatus) {
+  // Retrieve the list of topics for a given uE
```
[the remark this answers is not in the document]
Comment (mayureshkar) [reply up-core-api/uprotocol/core/udiscovery/v3/udiscovery.proto:56]: Added documentation for all APIs.

File: up-core-api/uprotocol/core/udiscovery/v3/udiscovery.proto [hunk 18]
```diff
@@ -139,233 +83,80 @@ service uDiscovery {
...
-    REMOVE = 3;     // Removed
-  }
-
-  // Time-to-live: How long (ms) the information should be live for in the database.
-  // A value of -1 means lives forever.
-  optional  int32 ttl = 4;
-
-  // uDiscovery resource that it serves (per SDV-202 definition): database
-  enum Resources { nodes = 0; }
+  v1.UUri UDomain = 1;
```
[the remark this answers is not in the document]
Comment (mayureshkar) [reply up-core-api/uprotocol/core/udiscovery/v3/udiscovery.proto:134]: changed

File: up-core-api/uprotocol/core/udiscovery/v3/udiscovery.proto [hunk 19]
```diff
@@ -139,233 +83,80 @@ service uDiscovery {
...
+// 2. code.NOT_FOUND: No matching UUris were found
+// 3. code.INVALID_ARGUMENT: The passed UUri is invalid
+message GetDeviceListResponse {
 
-// Observer Identification Information
-message ObserverInfo {
-  // Fully qualified URI for the Observer who is registering to receive the
-  // notifications from uDiscovery ex. `//vcu.VIN/com.gm.app.hartley`
-  string uri = 1;
+  repeated v1.UUri deviceList = 1;
```
[the remark this answers is not in the document]
Comment (mayureshkar) [reply up-core-api/uprotocol/core/udiscovery/v3/udiscovery.proto:144]: Changed

File: up-core-api/uprotocol/core/udiscovery/v3/udiscovery.proto [hunk 20]
```diff
@@ -139,233 +83,80 @@ service uDiscovery {
...
   v1.UUri uri = 1;                 // Resolved UUri  
-  uprotocol.v1.UStatus status = 3; // Return code for the rpc call
-}
-
-
-// Return value from LookupUri() API that contains the batch of Uris for the 
-// lookup and status from the API call
-message LookupUriResponse {
-  v1.UUriBatch uris = 1;           // Batch of Uris
   uprotocol.v1.UStatus status = 2; // Return code for the rpc call
```
[the remark this answers is not in the document]
Comment (mayureshkar) [reply up-core-api/uprotocol/core/udiscovery/v3/udiscovery.proto:209]: removed.

File: up-core-api/uprotocol/core/udiscovery/v3/udiscovery.proto [hunk 21]
```diff
@@ -58,74 +49,27 @@ service uDiscovery {
...
-  // **NOTE:** You MUST have write permission to the node who's property you
-  // are updating
-  rpc UpdateProperty(UpdatePropertyRequest) returns (uprotocol.v1.UStatus) {
-    option (method_id) = 7;
-  }
-
-
-  // Register to receive Notifications to changes to one of more Nodes. The changes
-  // are published on the notification topic: '/core.udiscovery/3/nodes#Notification'
-  rpc RegisterForNotifications(NotificationsRequest) returns (uprotocol.v1.UStatus) {
```
[the remark this answers is not in the document]
Comment (mayureshkar) [reply up-core-api/uprotocol/core/udiscovery/v3/udiscovery.proto:118]: Lets talk about this.

File: up-core-api/uprotocol/core/udiscovery/v3/udiscovery.proto [hunk 22]
```diff
@@ -139,233 +83,80 @@ service uDiscovery {
...
-
-  // Return code for the rpc call
-  uprotocol.v1.UStatus status = 2;
-
-  // Time-to-live: How long (ms) the information should be live for in the database.
-  // -1 means lives forever
-  optional int32 ttl = 3;
+// Return value from FinduE() API that contains the batch of Uris for the 
+// lookup and status from the API call
+message FinduEResponse {
```
[the remark this answers is not in the document]
Comment (mayureshkar) [reply up-core-api/uprotocol/core/udiscovery/v3/udiscovery.proto:147]: Lets talk about this.

File: up-core-api/uprotocol/core/udiscovery/v3/udiscovery.proto [hunk 23]
```diff
@@ -41,331 +41,300 @@ service uDiscovery {
...
-  // 1. code.OK: Query lookup was successful
-  // 2. code.NOT_FOUND: No matching UUris were found
-  // 3. code.INVALID_ARGUMENT: The passed UUri is invalid
-  // 4. code.PERMISSION_DENIED: The caller does not have permission to perform the query
-  rpc LookupUri(v1.UUri) returns (LookupUriResponse) {
-    option (method_id) = 1;
-  }
+  /* 
+  UEs shall call RegisterUE API to add themselves to discovery.
+ */
```
Comment (stevenhartley) [inline up-core-api/uprotocol/core/udiscovery/v3/udiscovery.proto:55]: Please provide more description for these APIs as I do not understand their purpose. Also we are revamping the UUri (please watch up-spec project to see what is changing).  comments in method definitions cannot have spaces between them for documentation generation purposes.

File: up-core-api/uprotocol/core/udiscovery/v3/udiscovery.proto [hunk 24]
```diff
@@ -41,331 +41,300 @@ service uDiscovery {
...
+  rpc RegisterUE(RegisterUERequest) returns ( uprotocol.v1.UStatus){
+    option (method_id) = 1;
   }
 
-  // Query the database to fetch a list of 1 or more properties for a given node.
-  rpc FindNodeProperties(FindNodePropertiesRequest) returns (FindNodePropertiesResponse) {
-    option (method_id) = 4;
-  }
+    /* 
+  UEs shall call UnregisterUE API to remove themselves from discovery.
```
Comment (stevenhartley) [inline up-core-api/uprotocol/core/udiscovery/v3/udiscovery.proto:67]: What does registering and unregistering do? Is this something that a uE actually does? How do you ensure that we don't call the API to spoof being someone else?

File: up-core-api/uprotocol/core/udiscovery/v3/udiscovery.proto [hunk 25]
```diff
@@ -41,331 +41,300 @@ service uDiscovery {
...
-  // **NOTE:** You MUST have write permission to the parent node
-  rpc AddNodes(AddNodesRequest) returns (uprotocol.v1.UStatus) {
-    option (method_id) = 6;
-  }
 
+  /* 
+  UEs shall call SetUETopics API to update discovery with the list of topics they service.
+  This will be used in two scenarios: 
+  1. when a UE is added on a device 
+  2. when UE wants to udpate the topic list
```
Comment (stevenhartley) [inline up-core-api/uprotocol/core/udiscovery/v3/udiscovery.proto:84]: Perhaps we register and pass the list of all the configured topics in one command.

File: up-core-api/uprotocol/core/udiscovery/v3/udiscovery.proto [hunk 26]
```diff
@@ -41,331 +41,300 @@ service uDiscovery {
...
-  // - code.INVALID_ARGUMENT: The passed UUri is invalid (missing names or numbers)
-  // - code.PERMISSION_DENIED: The caller does not have permission to perform the resolution
-  rpc ResolveUri(ResolveUriRequest) returns (ResolveUriResponse) {
-    option (method_id) = 10;
-  }
-}
 
+  /* 
+  UEs shall call SetUEProperties API to update discovery with updated property values. 
+  For example, this could be used to update an UE's version.
```
Comment (stevenhartley) [inline up-core-api/uprotocol/core/udiscovery/v3/udiscovery.proto:104]: properties are part of the UUri, what other properties are we setting? Perhaps we need an explicit structure like UServiceTopic but more a UServiceProperties.

File: up-core-api/uprotocol/core/udiscovery/v3/udiscovery.proto [hunk 27]
```diff
@@ -41,331 +41,300 @@ service uDiscovery {
...
+    option (method_id) = 4;
+  }
 
 
-// FindNodesRequest passed to FindNodes()
-message FindNodesRequest {
-  // uProtocol formatted URI for the node to search, ex. '//vcu.VIN/core.app.hartley'
-  // shall return the 'core.app.hartley' node
-  string uri = 1;
+ /*
```
Comment (stevenhartley) [inline up-core-api/uprotocol/core/udiscovery/v3/udiscovery.proto:172]: Why not call the GetUETopics() passing a wildcard?

File: up-core-api/uprotocol/core/udiscovery/v3/udiscovery.proto [hunk 28]
```diff
@@ -41,331 +41,300 @@ service uDiscovery {
...
-  uprotocol.v1.UStatus status = 2;
+  /******************************* Data replication / synchronization ******/
 
-  // Time-to-live: How long (ms) the information should be live for in the database.
-  // -1 means lives forever
-  optional int32 ttl = 3;
-}
+  /*
+    UpdateData is NOT a customer facing API. This shall be used discovery 
+    instances to push an update between to core components (streamer)/ other discovery instances.
```
Comment (stevenhartley) [inline up-core-api/uprotocol/core/udiscovery/v3/udiscovery.proto:194]: Then do not put this here. Also Streamer should not be updating information in the discovery database.

File: up-core-api/uprotocol/core/udiscovery/v3/udiscovery.proto [hunk 29]
```diff
@@ -41,331 +41,300 @@ service uDiscovery {
...
+  Example: 
+  UpdateData(“up:/domain/device/UE”)
+  Status: code.OK 
+  */
 
-// Find 1 or more properties for a given node passed to FindNodeProperties()
-message FindNodePropertiesRequest {
-  // the uri for the node in the database
-  string uri = 1;
+  rpc SyncData(UpdateDataRequest) returns (status){
```
Comment (stevenhartley) [inline up-core-api/uprotocol/core/udiscovery/v3/udiscovery.proto:203]: above it is called updateData and here it is called SyncData()

File: up-core-api/uprotocol/core/udiscovery/v3/udiscovery.proto [hunk 30]
```diff
@@ -41,331 +41,300 @@ service uDiscovery {
...
-// Returned from FindNodeProperties()
-message FindNodePropertiesResponse {
-  // a list of property name/value pairs
-  map <string, PropertyValue>  properties = 1;
-
-  // Return code for the rpc call
-  uprotocol.v1.UStatus status = 2;
-}
-
+/* All response messages shall include uprotocol status field. This shall indicate whether
```
Comment (stevenhartley) [inline up-core-api/uprotocol/core/udiscovery/v3/udiscovery.proto:210]: We are saying now in #112 that we do not want to also pass this in the reply but the RPC returns the status per commstatus (of response umessage)

**Review by stevenhartley — COMMENTED** [review]
Once I'm done the L1 and L2 definitions, I'll take a crack at this API to show you what I was thinking.

**Comment by stevenhartley** [comment]
Closing this pull request as it was replaced with #219
